Fix: export_published_chart correctly detects and exports tar file

svteb opened the pull request:
## Description
Removed file globbing as it caused errors and added a new function that will detect a specific tgz file to be untared. Removed the line: `FileUtils.rm_rf(tgz_name)` as it did not work previously and now that globbing has been fixed it would cause undesired behavior (honestly I was not able to figure out why it is there at all).

## Issues:
Refs: #1947 

## How has this been tested:
 - [x] Covered by existing integration testing
 - [ ] Added integration testing to cover
 - [ ] Verified all A/C passes
     * [ ] develop
     * [ ] master
     * [ ] tag/other branch
 - [x] Test environment
    * [ ] Shared Packet K8s cluster
    * [ ] New Packet K8s cluster
    * [ ] Kind cluster
    * [x] Minikube cluster
 - [ ] Have not tested

## Types of changes:
 - [x] Bug fix (non-breaking change which fixes an issue)
 - [ ] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
 - [ ] Documentation update

## Checklist:
**Documentation**
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] No updates required.

**Code Review**
- [x] Does the test handle fatal exceptions, ie. rescue block

**Issue**
- [x] Tasks in issue are checked off

martin-mat commented on src/tasks/utils/cnf_manager.cr:
```diff
@@ -1259,6 +1258,12 @@ module CNFManager
 
+  def self.get_tgz_name(helm_chart)
+    tar_files = Dir.glob("#{Helm.chart_name(helm_chart)}-*.tgz")
+    raise "No tar files found" if tar_files.empty?
```
more details about what filename was not found would make future troubleshooting easier.

svteb replied:
Resolved by adding a cleaner exception:

```
class TarFileNotFoundError < Exception
  def initialize(chart_name)
    super("No .tgz files found for chart #{chart_name}")
  end
end
```

martin-mat commented on src/tasks/utils/cnf_manager.cr:
```diff
@@ -1259,6 +1258,12 @@ module CNFManager
+  def self.get_tgz_name(helm_chart)
+    tar_files = Dir.glob("#{Helm.chart_name(helm_chart)}-*.tgz")
+    raise "No tar files found" if tar_files.empty?
+    tar_files.last
```
A warning in case that multiple files found?

svteb replied:
View the new change/read my new comment, this state is no longer possible.

martin-mat replied:
I checked the updated code but the point with possible multiple files is still here. At least issue a warning.

martin-mat commented on src/tasks/utils/cnf_manager.cr:
```diff
@@ -728,23 +728,22 @@ module CNFManager
     Log.info { "tgz_name: #{tgz_name}" }
 
     unless input_file && !input_file.empty?
-      FileUtils.rm_rf(tgz_name)
```
I guess the intention here was to remove an existing helm chart package to make sure it is pulled successfully?

svteb replied:
Yes, the intention was to remove all prior .tgz archives (did not work because this function is incompatible with file globbing). I backtracked on this removal and remade it so it would work.

svteb commented:
Comments from @martin-mat were correct in pointing out some unhandled cases/lack of information provided. I've come to understand the intentions behind `FileUtils.rm_rf(tgz_name)` and decided to bring it back:

```
unless input_file && !input_file.empty?
      files_to_delete = Dir.glob("#{Helm.chart_name(helm_chart)}-*.tgz") ~> addition
      files_to_delete.each do |file|
        FileUtils.rm(file)
        puts "Deleted: #{file}"
      end ~> end of addition

      helm_info = Helm.pull(helm_chart) 
      unless helm_info[:status].success?
        puts "Helm pull error".colorize(:red)
        raise "Helm pull error"
      end
end
```

It was necessary to do it like this because `FileUtils.rm_rf` does not work with file globbing which was originally present. This addition also removed the necessity of checking for state where multiple versions of some helm archive were present (as they get deleted and the newest one gets pulled subsequentially).

Another mistake from the previous commit, that was undiscovered, was the necessity of moving this entire block before the block where tgz archive is being discovered, as the `Helm.pull(helm_chart)` command is what downloads the tgz archive. This also makes the sequence of steps somewhat more logical:

`helm_pull -> discover_tgz -> extract_tgz` whereas the previous sequence was `discover_tgz -> helm_pull -> extract_tgz`.


Furthermore I added an exception that prints better information in case no tgz file is discovered.

martin-mat commented on src/tasks/utils/cnf_manager.cr:
```diff
@@ -1259,6 +1267,18 @@ module CNFManager
+    # In case no tgz file is present (should only happen if there was a prior helm failure).
+    if tgz_files.empty?
+      Log.error { "No .tgz files found for #{Helm.chart_name(helm_chart)}" }
+      raise TarFileNotFoundError.new(Helm.chart_name(helm_chart))
```
Better report as much information as we have now, like:
No files with glob "#{Helm.chart_name(helm_chart)}-*.tgz" found

martin-mat commented on src/tasks/utils/cnf_manager.cr:
```diff
@@ -1259,6 +1267,18 @@ module CNFManager
   end
 
+  def self.get_tgz_name(helm_chart)
+    tgz_files = Dir.glob("#{Helm.chart_name(helm_chart)}-*.tgz")
```
That glob pattern is used already twice so it's worth to have it on one place (a new helper method/getter?)

martin-mat replied:
comment still valid

martin-mat commented on src/tasks/utils/cnf_manager.cr:
```diff
@@ -1259,6 +1267,23 @@ module CNFManager
+      Log.error { "No .tgz files found for #{Helm.chart_name(helm_chart)}-*.tgz" }
+      raise TarFileNotFoundError.new(Helm.chart_name(helm_chart))
+    elsif tgz_files.size > 1
+      Log.info { "Multiple .tgz files found: #{tgz_files.join(", ")}" }
```
Log.warning
 plus note which one is being taken
?

svteb replied:
Later on in the code the picked tgz file is logged.

martin-mat commented on src/tasks/utils/cnf_manager.cr:
```diff
@@ -721,30 +720,39 @@ module CNFManager
       config = CNFManager::Config.parse_config_yml(CNFManager.ensure_cnf_testsuite_yml_path(config_file), airgapped: true) 
       tar_info = AirGap.tar_info_by_config_src(helm_chart)
       tgz_name = tar_info[:tar_name]
     elsif output_file && !output_file.empty?
```
simplify:
```
else 
      config = CNFManager::Config.parse_config_yml(CNFManager.ensure_cnf_testsuite_yml_path(config_file), generate_tar_mode: output_file && !output_file.empty?) 
      tgz_name = get_tgz_name(helm_chart)
end
```

svteb replied:
Good observation.

martin-mat commented on src/tasks/utils/cnf_manager.cr:
```diff
@@ -1259,6 +1267,27 @@ module CNFManager
+  end
+
+  def self.get_tgz_name(helm_chart)
+    tgz_files = tgz_files(helm_chart)
```
avoid same identifiers for different things

svteb replied:
Resolved

martin-mat commented on src/tasks/utils/cnf_manager.cr:
```diff
@@ -1259,6 +1267,27 @@ module CNFManager
     resource_keys.includes?(resource_key)
   end
 
+  def self.tgz_files(helm_chart)
```
be more explicit with method naming, for example
```
def self.find_tgz_files(helm_chart)
```

svteb replied:
Renamed

martin-mat commented on src/tasks/utils/cnf_manager.cr:
```diff
@@ -1259,6 +1267,27 @@ module CNFManager
+    Dir.glob(helm_tgz_expression(helm_chart))
+  end
+
+  def self.get_tgz_name(helm_chart)
```
be more descriptive/explicit with method naming, for example
```
def self.get_and_verify_tgz_name(helm_chart)
```

svteb replied:
Renamed

martin-mat commented on src/tasks/utils/cnf_manager.cr:
```diff
@@ -1259,6 +1267,27 @@ module CNFManager
+    tgz_files.first
+  end
+
+  def self.helm_tgz_expression(helm_chart)
```
better
```
def self.get_helm_tgz_expression(helm_chart)
```

svteb replied:
Renamed

martin-mat reviewed (approved):
lgtm, good refactoring!

kosstennbl commented on src/tasks/utils/cnf_manager.cr:
```diff
@@ -721,30 +720,36 @@ module CNFManager
+    # delete previous tgz files and pull the helm chart
     unless input_file && !input_file.empty?
-      FileUtils.rm_rf(tgz_name)
+      files_to_delete = find_tgz_files(helm_chart)
+      files_to_delete.each do |file|
+        FileUtils.rm(file)
+        Log.info { "Deleted: #{file}" }
+      end
+
       helm_info = Helm.pull(helm_chart) 
       unless helm_info[:status].success?
         puts "Helm pull error".colorize(:red)
         raise "Helm pull error"
       end
     end
 
-    TarClient.untar(tgz_name,  "#{destination_cnf_dir}/exported_chart")
+    # get the tgz file name
+    if input_file && !input_file.empty?
+      # todo add generate and set tar as well
+      config = CNFManager::Config.parse_config_yml(CNFManager.ensure_cnf_testsuite_yml_path(config_file), airgapped: true) 
+      tar_info = AirGap.tar_info_by_config_src(helm_chart)
+      tgz_name = tar_info[:tar_name]
+    else
+      config = CNFManager::Config.parse_config_yml(CNFManager.ensure_cnf_testsuite_yml_path(config_file), generate_tar_mode: output_file && !output_file.empty?) 
+      tgz_name = get_and_verify_tgz_name(helm_chart)
+    end
```
I think that these "if" statements should be joined and each logical branch marked with comments:
```
if input_file && !input_file.empty?   # Input file present, Tar installation mode
    ...
else   # Input file absent, Pulled chart installation mode
    ...
end
```

svteb replied:
Great catch, I reduced it to one if else block, should be alright now.

kosstennbl reviewed (commented):
Other than mentioned problem - looks nice: fixes the Tar problem and refactors the code
